wallet-new/src/Cardano/Wallet/API/V1/LegacyHandlers/Info.hs

```diff
@@ -53,3 +58,9 @@ getInfo Diffusion{..} ntpStatus ntpCheck = do
             { timeDifferenceFromNtpServer = fmap V1.mkLocalTimeDifference timeDifference
             }
         }
+  where
+    checkNtpBlocking = do
+        s <- readTVar ntpStatus
+        case s of
+            NtpSyncPending -> checkNtpBlocking
+            state          -> newTVar state
```

Comment:
I don't think this is what we want to do, but I'm not :100: on STM semantics. The way this code is written, we are going to:

1. Start a transaction with `atomically`. 
2. Write `NtpSyncPending` to the `TVar`.
3. Read the status. 
    a. If it is not `NtpSyncPending`, then we return a new TVar with the value. This ends the `atomically` block which commits the transaction.
    b. If it is `NtpSyncPending`, then we loop.

However -- this is an atomic transaction! Which means that any `write` to the TVar is going to cause our `read` to go stale, which is going to cause it to `retry` the entire transaction, including the `writeTVar ntpStatus NtpSyncPending` call. If I am right, then this will loop forever.

An implementation that avoids this will look like:

```haskell
atomically $ writeTVar ntpStatus NtpSyncPending
let loop = do
        var <- atomically $ readTVar ntpStatus
        case var of
            NtpSyncPending -> threadDelay 1000 >> loop
            _ -> pure var
loop
```

Reply:
Indeed this will not work.  The problem is that the ntp client will not have a chance to notice that the state has changed (since everything runs in a single stm transaction). By slightly improving @personsmatt snippet, I suggest we do the following:
```
-- wake up ntp client
atomically $ writeTVar ntpStatus NtpSyncPending
-- wait until status changes and get the new value returned by the client 
status <- atomically $ do
        var <- readTVar ntpStatus
        case var of
            NtpSyncPending -> retry
            _ -> pure var
```
The advantage is that we are not polling, but the stm runtime will wait until the client will update the `TVar`.

Reply:
Wow, is it really that smart? My understanding of this code is that you'll have a thread spinning on the TVar constantly.

Reply:
Yes it is, that's why it's so nice! It's not optimized though, so you might want to read check if the `TVar` has changed and write only if you the value differs. @parsonsmatt your thread would poll, mine would be woken up the STM runtime.

Reply:
I love it when Haskell is even smarter than I thought :joy: